forte/data/vocabulary.py

```diff
@@ -0,0 +1,117 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=line-too-long
+from typing import List, Tuple, Dict, Union, Hashable, Iterable
+
+
+class Vocabulary:
+    """This class will store "Elements" that are added, assign "Ids" to them and
+    return "Representations" if queried. These three are the main concepts in this class.
+
+    1. Element: Any hashable instance that the user want to store.
+    2. Id: Each element will have an unique Id, which is an interger.
+    3. Representation: according to the configuration, the representation for an element
+        could be an interger (in this case, would be "Id"), or an one-hot vector (in this
+        case, would be a list of interger).
+
+    There are two special elements.
+
+    1. One is <PAD> element, which will be mapped into Id of 0 or -1 and have different
+        representation according to different setting.
+    2. The other one is <UNK> element, which, if added into the vocabulary, will be the
+        default element if the queried element is not found.
+
+    Here is a table on how our Vocabulary class behavior under different settings.
+
+    User_request = {
+        "vocab method":  "raw"(handle outside), "indexing",      "indexing",          "one-hot",                    "one-hot"
+        "need_pad":           assume False,        True,            False,              True,                         False
+    }
+
+    Vocabulary behavior:
+        get_pad_value            None                0              None                [0,0,0]                        None
+        inner_mapping            None           0:pad  1:ele0      0:ele0         -1:<PAD>   0:ele0                   0:ele0
+        element2repr           raise Error      pad->0 ele0->1     ele0->0      <PAD>->[0,0,0] ele0->[1,0,0]        ele0->[1,0,0]
+        id2element             raise Error      0->pad 1->ele0     0->ele0    -1 -> <PAD> 0->ele0 (be careful)        0->ele0
+
+    Not implemented for simplicity:
+        element2id             raise Error      pad->0 ele0->1     ele0->0        <PAD> -> -1   ele0 -> 0             ele0->0
+        repr2element           raise Error,     0->pad 1->ele0     0->ele0     [0,0,0]-><PAD> [1,0,0]->ele0        [1,0,0]->ele0
+    """
+    PAD_ELEMENT = "<PAD>"
+    UNK_ELEMENT = "<UNK>"
+
+    def __init__(self, method: str, need_pad: bool, use_unk: bool):
+        self.method = method
+        self.need_pad = need_pad
+        self.use_unk = use_unk
+
+        self.element2id_dict: Dict = dict()
+        self.id2element_dict: Dict = dict()
+
+        if method == "one-hot" and need_pad:
+            self.next_id = -1
+        else:
+            self.next_id = 0
+
+        if need_pad:
+            self.add_element(Vocabulary.PAD_ELEMENT)
+
+        if use_unk:
+            self.add_element(Vocabulary.UNK_ELEMENT)
+
```

Comment:
The public functions below require docstring. If you do not intend to make them public, add `_`

Reply:
Got it.